File: verbNet/src/main/java/org/sqlunet/verbnet/sql/VnEntry.java
/*
 * Copyright (c) 2019. Bernard Bou <1313ou@gmail.com>.
 */

package org.sqlunet.verbnet.sql;

import android.database.sqlite.SQLiteDatabase;

import org.sqlunet.wordnet.sql.BasicWord;

import java.util.ArrayList;
import java.util.List;

import androidx.annotation.Nullable;

/**
 * VerbNet entry
 *
 * @author <a href="mailto:1313ou@gmail.com">Bernard Bou</a>
 */
public class VnEntry
{
	/**
	 * Word for this entry
	 */
	public final BasicWord word;

	/**
	 * Synsets attached to entry
	 */
	public final List<VnSynset> synsets;

	/**
	 * Constructor
	 *
	 * @param word    word string
	 * @param synsets list of synsets attached to this entry
	 */
	private VnEntry(final BasicWord word, final List<VnSynset> synsets)
	{
		this.word = word;
		this.synsets = synsets;
	}

	/**
	 * Make entry
	 *
	 * @param connection connection
	 * @param word       target word
	 * @return new entry or null
	 */
	@Nullable
	static public VnEntry make(final SQLiteDatabase connection, final String word)
	{
		VnEntry entry = null;
		VnClassQueryFromWordAndPos query = null;
		try
		{
			query = new VnClassQueryFromWordAndPos(connection, word);
			query.execute();

			long wordId = 0;
			List<VnSynset> synsets = null;

			while (query.next())
			{
				wordId = query.getWordId();
				if (synsets == null)
				{
					synsets = new ArrayList<>();
				}
				synsets.add(new VnSynset(query));
			}
			if (wordId != 0 && synsets != null)
			{
				entry = new VnEntry(new BasicWord(word, wordId), synsets);
			}
		}
		finally
		{
			if (query != null)
			{
				query.release();
			}
		}
		return entry;
	}
}

File: verbNet/src/main/java/org/sqlunet/verbnet/sql/VnEntry.kt
/*
 * Copyright (c) 2023. Bernard Bou
 */
package org.sqlunet.verbnet.sql

import android.database.sqlite.SQLiteDatabase
import org.sqlunet.wordnet.sql.BasicWord

/**
 * VerbNet entry
 *
 * @param word    word string
 * @param synsets list of synsets attached to this entry
 *
 * @author [Bernard Bou](mailto:1313ou@gmail.com)
 */
class VnEntry private constructor(
    val word: BasicWord,
    val synsets: List<VnSynset>?,
) {

    companion object {

        /**
         * Make entry
         *
         * @param connection connection
         * @param word       target word
         * @return new entry or null
         */
        fun make(connection: SQLiteDatabase, word: String): VnEntry? {
            VnClassQueryFromWordAndPos(connection, word).use {
                it.execute()
                var wordId: Long = 0
                var synsets: MutableList<VnSynset>? = null
                while (it.next()) {
                    wordId = it.wordId
                    if (synsets == null) {
                        synsets = ArrayList()
                    }
                    synsets.add(VnSynset(it))
                }
                if (wordId != 0L) {
                    return VnEntry(BasicWord(word, wordId), synsets)
                }
            }
            return null
        }
    }
}
